---
title: "ERCC Tool Dependencies"
author: "James Diao"
date: "17 February 2017"
output:
  html_document:
    code_folding: hide
    number_sections: yes
    toc: yes
  pdf_document:
    number_sections: yes
    toc: yes
geometry: margin=1in
---

**Working Directory**: `r getwd()`

```{r setup, include = F}
# change echo = T for code-folding in HTML document
### Used for troubleshooting
setwd("/Users/jamesdiao/Documents/Gerstein/CBB_Project_2016")
knitr::opts_chunk$set(echo = T, eval = T, cache = F, warning = F, message = F)
```

```{r packages, include = F, cache = F}
# Install and load required packages
pkg_list <- c("knitr","pander","ggplot2","tsne","tibble","curl","tidyr","tsne","dplyr")
installed <- pkg_list %in% installed.packages()[,"Package"]
if (!all(installed))
  install.packages(pkg_list[!installed])
sapply(pkg_list, require, character.only = T)
```

```{r import_tsv} 
### This code chunk is performed in the actual app, but we also need to do it here 
### to generate the other dependencies. 
### Data_Summary_1567.tsv was downloaded from http://exrna-atlas.org/exat/gridview using tabletools2.
import_tsv <- function(data_file) {
  file.by.line <- suppressWarnings(readLines(data_file) %>% strsplit("\t"))
  data_summary <- do.call(rbind, file.by.line) %>% as.data.frame(stringsAsFactors = F)
  if (ncol(data_summary) == 19) {
    colnames(data_summary) <- c("Blank","Biosample_Name","Condition",
      "Anatomical_Location","Biofluid_Name","exRNA_Source","Cell_Culture_Source",
      "Profiling_Assay","RNA_Isolation_Kit","ERCC_QC_Meets_Standards",
      "ERCC_QC_Reference_Genome_Reads","ERCC_QC_Transcriptome_Genome_Ratio",
      "ERCC_QC_Transcriptome_Reads","Download_Data","Download_Advanced_Results",
      "Download_Metadata","RNA_Profile","External_References","Biosample_Metadata_Accession")
  } else {
    colnames(data_summary) <- c("Blank","Biosample_Name","Condition","Anatomical_Location",
      "Biofluid_Name","exRNA_Source","Cell_Culture_Source",
      "Profiling_Assay","RNA_Isolation_Kit","ERCC_QC_Meets_Standards",
      "ERCC_QC_Reference_Genome_Reads","ERCC_QC_Transcriptome_Genome_Ratio",
      "ERCC_QC_Transcriptome_Reads","Download_Data","Download_Metadata",
      "External_References","Biosample_Metadata_Accession")
  }
  data_summary <- data_summary[,apply(data_summary, 2, function(col) any(nzchar(col)))]
  data_summary <- suppressWarnings(data_summary %>% 
    mutate(ERCC_QC_Reference_Genome_Reads = gsub(",","",ERCC_QC_Reference_Genome_Reads) %>% as.integer) %>%
    mutate(ERCC_QC_Transcriptome_Reads = gsub(",","",ERCC_QC_Transcriptome_Reads) %>% as.integer) %>%
    mutate(ERCC_QC_Transcriptome_Genome_Ratio = ERCC_QC_Transcriptome_Genome_Ratio %>% as.numeric) %>% 
    mutate(Biofluid_Name = sub("Cerebrospinal fluid","CSF", 
                           sub("Culture Media, Conditioned","Cultured_Media", Biofluid_Name))))
  biofluid_names <- table(data_summary$Biofluid_Name) %>% sort(decreasing = T) %>% names
  data_summary <- mutate(data_summary, Biofluid_Name = factor(Biofluid_Name, levels = biofluid_names))
}
data_summary <- import_tsv("Data_Summary_1567.tsv") %>% arrange(Biosample_Metadata_Accession)
#%>% filter(ERCC_QC_Meets_Standards == "YES")
```

```{r rna_reads} 
### This code chunk gives rna_reads -> log_rna_reads, which is then used for PCA and tSNE
collect_rna <- function(prefix) {
  # Collect file names
  RPM_files <- system(sprintf("ls %s", prefix), intern = T) 
  datasets <- RPM_files %>% strsplit("-2016") %>% sapply("[[",1)
  # Removing outlier: "TPATE1-human-plasma-healthyVsCancer"
  RPM_files <- RPM_files[!grepl("TPATE1-human-plasma-healthyVsCancer",RPM_files)]
  datasets <- datasets[!grepl("TPATE1-human-plasma-healthyVsCancer",datasets)] 
  # Import
  if (prefix == "smallRNAQuants") {
    rna_all <- lapply(RPM_files, function(data) {
      load(file = sprintf("smallRNAQuants/%s", data))
      exprs.gencode.rpm[grep("snRNA",rownames(exprs.gencode.rpm)),]
    }) %>% setNames(datasets)
    prefix <- "snRNA"
  } else {
    rna_all <- lapply(RPM_files, function(data) {
      read.table(file = sprintf("%s/%s", prefix,data), header = TRUE)
    }) %>% setNames(datasets)
  }
  # Collect #miRNAs and #samples for each dataset
  rna_dim <- sapply(rna_all, function(x) 
    dim(x) %>% setNames(c(sprintf("%ss", prefix),"Samples"))) %>% t
  rna_dim
  if (prefix == "miRNA") {
    # Vector of dataset names
    sample_map <- rownames(rna_dim)[lapply(1:nrow(rna_dim), function(x) rep(x,rna_dim[x,2])) %>% unlist]
    saveRDS(sample_map, file = "Dependencies/sample_map.rds")
  }
  # Names of all miRNAs
  rna_names <- lapply(rna_all, rownames) %>% unlist %>% unique
  # Joining across datasets
  do.call("cbind", lapply(rna_all, function(rna){
    expanded <- matrix(0, nrow = length(rna_names), ncol = ncol(rna)) %>% as.data.frame
    rownames(expanded) <- rna_names
    colnames(expanded) <- colnames(rna)
    expanded[rownames(rna),] <- rna
    return(expanded)
  })) %>% t -> rna_reads
  # Standardize naming
  rownames(rna_reads) <- gsub("\\.","-",rownames(rna_reads))
  # Order and log-transform
  rna_reads <- rna_reads[order(rownames(rna_reads)),]
  return(rna_reads)
}

small_num <- function(reads) {
  10^(floor(log10(min(reads[reads!=0]))))
}

rna_reads <- lapply(c("miRNA","piRNA","tRNA","smallRNAQuants"), collect_rna) %>% 
  setNames(c("miRNA","piRNA","tRNA","snRNA"))
log_rna_reads <- lapply(rna_reads, function(rna) log10(rna+small_num(rna)))
```

```{r sample_map} 
### This code chunk allows us to join information from Data_Summary_1567 and rna_reads
QC_files <- system("ls QC_Results", intern = T) 
datasets <- QC_files %>% strsplit("-2016") %>% sapply("[[",1)
# Removing outlier: "TPATE1-human-plasma-healthyVsCancer"
QC_files <- QC_files[!grepl("TPATE1-human-plasma-healthyVsCancer",QC_files)]
datasets <- datasets[!grepl("TPATE1-human-plasma-healthyVsCancer",datasets)] 
# QC_list: list of all QC values in each of 10 datasets
QC_list <- lapply(QC_files, function(data) {
  read.table(file = paste0("QC_Results/",data), header = TRUE)
}) %>% setNames(datasets)
# QC_all: combines QC_list into a single data frame
QC_all <- do.call("rbind",QC_list) 
QC_all <- data.frame(QC_all, Name = rownames(QC_all))
QC_all <- select(QC_all, Name, everything())
QC_all <- arrange(QC_all, Name)
QC_all$Name <- gsub("\\.","-",QC_all$Name)

map <- sapply(1:nrow(QC_all), function(row_num) {
  # temp: compares 1 line of rna_reads_QC to data_summary_QC, list of "hits". Max hits is assigned. 
  temp <- c( which(QC_all[row_num, "TranscriptomeReads"] == data_summary$ERCC_QC_Transcriptome_Reads),
           which(QC_all[row_num, "TranscriptomeGenomeRatio"] == data_summary$ERCC_QC_Transcriptome_Genome_Ratio),
           which(QC_all[row_num, "GenomeReads"] == data_summary$ERCC_QC_Reference_Genome_Reads)
          ) 
  temp[which.max(table(temp))]
})
saveRDS(map, file = "Dependencies/map.rds")
```

```{r pca_plots}
### Generate PCA embedding
log_reads_pca <- lapply(log_rna_reads, function(reads) {
  rna_pca <- prcomp(reads, center = T, scale. = F)
  reduced_cols <- seq(1,min(100,ncol(rna_pca$x)))
  rna_pca$x[,reduced_cols]
})
saveRDS(log_reads_pca, file = "Dependencies/all_log_PCA_reduced.rds")
```

```{r tsne_plots}
### Generate tSNE embedding
### WARNING: THIS WILL TAKE A LONG TIME. To reduce runtime, use max_iter = 100. 
log_reads_tsne <- lapply(log_rna_reads, function(reads) {
  log_tsne_out <- tsne(reads, k=2, initial_dims = 30, perplexity = 30, max_iter = 1000, epoch = 50)
  return(data.frame("tSNE_1" = log_tsne_out[,1], "tSNE_2" = log_tsne_out[,2]))
})
saveRDS(log_reads_tsne, file = "Dependencies/all_log_tSNE_reduced.rds")
```


